### PR TITLE
ImportC: support dllimport, dllexport, naked attributes

### PR DESCRIPTION
**Pull request by WalterBright**
Of interest mainly on Windows

### CODE REVIEW

**Comment by WalterBright** [comment]
No idea why DAutoTest is failing

File: compiler/src/dmd/func.d [hunk 1]
```diff
@@ -214,11 +215,14 @@ private struct FUNCFLAG
     bool hasNoEH;            /// No exception unwinding is needed
     bool inferRetType;       /// Return type is to be inferred
     bool hasDualContext;     /// has a dual-context 'this' parameter
+
```
Comment (dkorpel) [inline compiler/src/dmd/func.d:218]: Newline looks unintentional

**Review by dkorpel — APPROVED** [review]

File: compiler/src/dmd/func.d [hunk 2]
```diff
@@ -204,6 +204,7 @@ private struct FUNCFLAG
     bool hasCatches;         /// function has try-catch statements
     bool skipCodegen;        /// do not generate code for this function.
     bool printf;             /// is a printf-like function
+
```
[the remark this answers is not in the document]
Comment (WalterBright) [reply compiler/src/dmd/func.d:207]: It's intentional. I wanted to group them by 10s so it's easy to see if they all fit in an int.

File: compiler/src/dmd/frontend.h [hunk 3]
```diff
@@ -2525,6 +2525,10 @@ class FuncDeclaration : public Declaration
     bool hasEscapingSiblings(bool v);
     bool computedEscapingSiblings() const;
     bool computedEscapingSiblings(bool v);
+    bool dllImport() const;
```
[the remark this answers is not in the document]
Comment (WalterBright) [reply compiler/src/dmd/frontend.h:2528]: ok

File: compiler/src/dmd/func.d (hunk 1 above)
Comment (WalterBright) [reply compiler/src/dmd/func.d:218]: ditto

File: compiler/test/compilable/cattributes.i [hunk 4]
```diff
@@ -0,0 +1,31 @@
+/* Smoke test dllimport, dllexport, and naked attributes */
```
[the remark this answers is not in the document]
Comment (dkorpel) [reply compiler/test/compilable/cattributes.i:1]: That way regressions won't be caught though.

File: compiler/src/dmd/cparse.d [hunk 5]
```diff
@@ -4880,6 +4886,31 @@ final class CParser(AST) : Parser!AST
...
+     * Add attributes from Specifier to variable
+     * Params:
+     *  vd = function to apply them to
+     *  specifier = specifiers
+     */
+    void specifiersToVarDeclaration(AST.VarDeclaration vd, const ref Specifier specifier)
+    {
+        vd.dllImport = specifier.dllimport;
+        vd.dllExport = specifier.dllexport;
+    }
```
Comment (ibuclaw) [inline compiler/src/dmd/cparse.d:4912]: Just wrap it in a VisibilityDeclaration, no reason to add needless duplicate information.

**Review by ibuclaw — DISMISSED** [review]

File: compiler/src/dmd/cparse.d (hunk 5 above)
Comment (WalterBright) [reply compiler/src/dmd/cparse.d:4912]: VisibilityDeclaration doesn't deal with dllimport/dllexport. If it was added, then it still would have to be propagated from the VisibilityDeclaration to the var/func declaration during the semantic pass. The duplication will still be there. I don't think this is a win.

The point to the VisibilityDeclaration was to enable the block scope of such attributes. This is irrelevant for dllimport/dllexport, which never apply to anything but the specific declaration the are associated with.

You could argue that it should set the Visibility export bit, but that has slightly different semantics than dllexport/dllimport, i.e. there is some impedance mismatch.

**Pull request merged**